src/backend/catalog/dependency.c

```diff
@@ -1351,7 +1351,8 @@ deleteOneObject(const ObjectAddress *object, Relation *depRel, int flags)
 
 	while (HeapTupleIsValid(tup = systable_getnext(scan)))
 	{
-		CatalogTupleDelete(*depRel, &tup->t_self);
+		if (!ENRdropTuple(*depRel, tup))
```

Comment:
There is a problem with the way systable_getnext handles ENR while deleting something from the list.

```
520     if (sysscan->enr)
521     {
...
525         if (sysscan->enr_tuplist && sysscan->enr_tuplist_i < sysscan->enr_tuplist->length)
526         {
527             htup = lfirst(list_nth_cell(sysscan->enr_tuplist, sysscan->enr_tuplist_i++));
528         }
529     }
```

Consider the case when enr->tuplist = [1, 2, 3, 4, 5, 6, 7, 8], ie length = 8 and enr_tup_list_i = 0
Now we call line 527, so it returns item at index 0 and it does enr_tup_list_i++.

Next we call ENRdropTuple to remove the item from the list.
Now list becomes [2, 3, 4, 5,  6, 7,  8] and enr_tup_list_i = 1 so next iteration of the while loop will return item at index 1 which is 3.

I think we should do this instead

```
if (!scan->enr) {
   while (HeapTupleIsValid(tup = systable_getnext(scan)))
       CatalogTupleDelete(*depRel, &tup->t_self);
} else {
   // drop everything from scan->enr_tuplist
}
```

Reply:
@surendravishnoi is correct. I hit one crash with this. `deleteOneObject` caller is recursive so the object we're trying to delete here may already been drop so `ENRdropTuple` would return false here so we would call `CatalogTupleDelete` which would crash because tuple is not in catalog.

We can do this instead

```
if (scan->enr)
            ENRdropTuple(*depRel, tup);
else
            CatalogTupleDelete(*depRel, &tup->t_self);
```

Reply:
I've fixed this in https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/130